### DISCUSSION TITLE
Empty home feed after reboot - should redis be persisting? or are they supposed to regenerate?

### DISCUSSION THREAD

**Discussion by alexbbrown** [mastodon/mastodon#22960] [General]
Hi folks, new server admin here.

When my mastodon server restarts, due to peril or maintenance, my home timeline on my test user becomes empty.  

<img width="315" alt="image" src="https://user-images.githubusercontent.com/2532344/210973115-0f0814de-75ad-4a87-8b7f-ad1f13cfc5e3.png">

Refreshing doesn't help, and waiting a bit doesn't help.

I can fix this by running 
```bash
RAILS_ENV=production bundle exe tootctl feeds build
```
Then all the posts of people I'm following appear again.

## What should happen?

So my question is, what's expected experience here?  and what's the expected underlying behaviour

I think the expected user experience is that the *home feed should just be present when the user wants it*

This should presumably be arranged by one of

* the feed persists over restart and there isn't a problem
* the feed is manually rebuilt after restart
* the feed is automatically rebuilt when the user tries to look at it
* the feed is automatically rebuilt when the user tries to look at it
* OR: the feed is expected to be empty.  tough, don't reboot

Does anyone know which of these is more likely to be true?

Does my set-up have a bug in it?  
it's vanilla 4.0.2, wrapped in some scripts called 'supervisord'.

### evidence I have found

* restarting redis loses users feed
* refreshing an empty feed doesn't fix it
* waiting doesn't fix it
* logging into shell and running `tootctl feed rebuild` fixes it (until next time)
* telling redis to `BGSAVE` and *then* restarting redis does not lose the feeds.  This seems most promising but at what cost?

## why this is important to me

I'm running on a shared host, and periodically they terminate my server for memory overruns.  Sometimes I tweak config and run restart.  each time my home feed becomes empty and boring, and I assume that the home feeds of my users do too.

**Comment by alexbbrown** [comment]
This is based on the default configuration provided by OpalStack.com

I have one redis configured

**Comment by themullinator** [comment]
This sounds _approximately_ like expected behavior in the sense that Mastodon does not recommend persisting the Redis cache (https://docs.joinmastodon.org/admin/scaling/#redis), although interestingly I have a default non-persistent Redis configuration on my instance and don't have to manually trigger a timeline rebuild even after a hard reboot of the host. Are there any errors in your redis logs?

**Reply by alexbbrown** [reply]
I interpret that paragraph differently; that it's only the LRU cache at `CACHE_REDIS_URL` that you should not persist.

I think there's an unwritten assumption in the admin docs that the main redis store at `REDIS_URL` is persisted.  I'm not running Ubuntu so I can't check what the default config for redis is.

I found this 3rd party guide to configuring sidekiq and redis that helps:
https://github.com/mperham/sidekiq/wiki/Using-Redis#multiple-redis-instances
> Many people use [Redis as a cache](http://redis.io/topics/lru-cache) (it works great as a [Rails cache store](http://guides.rubyonrails.org/caching_with_rails.html#activesupport-cache-rediscachestore)) but it's important that Sidekiq be run against a Redis instance that is not configured as a cache but as a persistent store. I recommend using two separate Redis instances, each configured appropriately, if you wish to use Redis for caching and Sidekiq. Redis namespaces do not allow for this configuration and come with [many other problems](http://www.mikeperham.com/2015/09/24/storing-data-with-redis/), so using discrete Redis instances is always preferred.

---
I found that article via this other article on tuning mastodon: https://hazelweakly.me/blog/scaling-mastodon/

**Comment by jmrplens** [comment]
I solved with systemd directive `Requires=your-redis-mastodon.service` in every mastodon service.

I actually have it configured with `slices` (settings tuned for tests purposes):

```bash
mastodon-web.slice - Mastodon Web Slice
     Loaded: loaded (/etc/systemd/system/mastodon-web.slice; enabled; vendor preset: enabled)
     Active: active since Wed 2023-08-09 12:03:38 CEST; 7min ago
     CGroup: /mastodon.slice/mastodon-web.slice
             ├─mastodon-web-1.service
             │ ├─342776 puma 5.6.5 (tcp://127.0.0.1:3001) [live]
             │ ├─344582 puma: cluster worker 0: 342776 [live]
             ├─mastodon-web-2.service
             │ ├─342778 puma 5.6.5 (tcp://127.0.0.1:3002) [live]
             │ ├─345201 puma: cluster worker 0: 342778 [live]
             ├─mastodon-web-3.service
             │ ├─342779 puma 5.6.5 (tcp://127.0.0.1:3003) [live]
             │ ├─344317 puma: cluster worker 0: 342779 [live]
             └─mastodon-web.service
               ├─342780 puma 5.6.5 (tcp://127.0.0.1:3000) [live]
               ├─345016 puma: cluster worker 0: 342780 [live]

 mastodon-streaming.slice - Mastodon Streaming Slice
     Loaded: loaded (/etc/systemd/system/mastodon-streaming.slice; enabled; vendor preset: enabled)
     Active: active since Wed 2023-08-09 12:03:38 CEST; 8min ago
     CGroup: /mastodon.slice/mastodon-streaming.slice
             ├─mastodon-streaming-1.service
             │ ├─342735 /usr/bin/node ./streaming
             │ ├─343159 /usr/bin/node /home/mastodon/live/streaming
             ├─mastodon-streaming-2.service
             │ ├─342737 /usr/bin/node ./streaming
             │ ├─343110 /usr/bin/node /home/mastodon/live/streaming
             ├─mastodon-streaming-3.service
             │ ├─342738 /usr/bin/node ./streaming
             │ ├─343082 /usr/bin/node /home/mastodon/live/streaming
             └─mastodon-streaming.service
               ├─342739 /usr/bin/node ./streaming
               ├─343111 /usr/bin/node /home/mastodon/live/streaming

 mastodon-sidekiq.slice - Mastodon Sidekiq Slice
     Loaded: loaded (/etc/systemd/system/mastodon-sidekiq.slice; enabled; vendor preset: enabled)
     Active: active since Wed 2023-08-09 12:03:39 CEST; 9min ago
     CGroup: /mastodon.slice/mastodon-sidekiq.slice
             ├─mastodon-sidekiq-default-1.service
             │ └─342873 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-default.service
             │ └─342874 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-ingress-1.service
             │ └─342875 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-ingress.service
             │ └─342876 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-mailer.service
             │ └─342877 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-pull-1.service
             │ └─342878 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-pull-2.service
             │ └─342879 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-pull.service
             │ └─342880 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-push-1.service
             │ └─342881 sidekiq 6.5.8 live [0 of 40 busy]
             ├─mastodon-sidekiq-push.service
             │ └─342882 sidekiq 6.5.8 live [0 of 40 busy]
             └─mastodon-sidekiq-scheduler.service
               └─342883 sidekiq 6.5.8 live [0 of 40 busy]
```

Content of the three slices:

```ini
[Unit]
Description=Mastodon XXX Slice
Before=slices.target 
Requires=mastodon-redis.service mastodon-redis-cache.service
```

Redis services:

```bash
mastodon-redis.service - Advanced key-value store
     Loaded: loaded (/etc/systemd/system/mastodon-redis.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2023-08-09 12:03:38 CEST; 11min ago
   Main PID: 342709 (redis-server)
     Status: "Ready to accept connections"
      Tasks: 7 (limit: 35973)
     Memory: 42.3M
        CPU: 4min 21.062s
     CGroup: /system.slice/mastodon-redis.service
             └─342709 /usr/bin/redis-server 127.0.0.1:6380

mastodon-redis-cache.service - Advanced key-value store
     Loaded: loaded (/etc/systemd/system/mastodon-redis-cache.service; enabled; vendor preset: enabled)
     Active: active (running) since Wed 2023-08-09 12:03:38 CEST; 12min ago
   Main PID: 342705 (redis-server)
     Status: "Ready to accept connections"
      Tasks: 6 (limit: 35973)
     Memory: 9.8M
        CPU: 1.919s
     CGroup: /system.slice/mastodon-redis-cache.service
             └─342705 /usr/bin/redis-server 127.0.0.1:6381
```
